### DISCUSSION TITLE
Harvard CS50: This is CS50x - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#87] [Announcements]
# Harvard CS50: This is CS50x - CS自学指南

CS自学指南

https://csdiy.wiki/%E7%BC%96%E7%A8%8B%E5%85%A5%E9%97%A8/CS50/

**Comment by coyoteshkw** [comment]
今年这个Codespace挂上魔法也创建不了:( ，去年的可以进去，今年的无论怎么尝试都是「codespace creation failed」

**Reply by Emanonstar** [reply]
实测有梯子是可以创建的

**Reply by coyoteshkw** [reply]
我之后重新注册了一个账号又可以了，不过好久没弄了，现在估计又忘记了😂

**Reply by TheLittleLightBulb** [reply]
挂着梯子进入为什么显示codespace fail to connect😂

**Reply by Manchesterblu** [reply]
我人在国外都创建不了codespace....也是codespace creation failed

**Reply by Chocolatieee0929** [reply]
有哪位大佬知道check50 出现的You seem to be missing these required files错误吗，相应的文件夹里明明就包含了对应的py文件，找了好久都没解决好T^T
```
$ check50 cs50/problems/2022/python/scourgify
Connecting......
You seem to be missing these required files:
scourgify.py
You are currently in: /workspaces/115781945, did you perhaps intend another directory?
$ ls cs50/problems/2022/python/scourgify
scourgify.py
```

**Reply by Chocolatieee0929** [reply]
我已经找到解决方法了，需要cd对应的文件夹里check，例如cd cs50/problems/2022/python/scourgify，再进行check50 cs50/problems/2022/python/scourgify

**Reply by liweisu59** [reply]
我这边是上了梯子的，但是是ECONNABORTED: (timeout) timeout of 20000ms exceeded，记得去年还能上的呀（那时在学校里

**Comment by mancuoj** [comment]
https://github.com/mancuoj/CS50x 在github更新了自己关于习题集的解答，求个star

**Reply by Andy-xiaokang** [reply]
感谢分享这门课程，这门课程质量真的高，c语言好多细节的位置讲的真清晰，短短五节课的C语言却让人茅塞顿开，解释了好多困惑

**Reply by figuretu** [reply]
我正在进行这门课的翻译，计划是 C 语言部分翻译 23 年版本，后面部分翻译 24 年新版。
同时，因为初学者的问题大多相似，维护一个共享文档收录有价值的讨论和学习资料：https://uufyjevghz.feishu.cn/docx/DP78d2U5TosTOTx9QCbcjp8GnBh
交流群目前 300 人左右，群号在文档里。入群需要看完 Lecture 0，完成验证问题。

**Reply by Rimor4** [reply]
请问交流群是在哪个平台呢，qq微信都搜不到

**Reply by ccckkk567** [reply]
就是qq吧

**Reply by Chononut** [reply]
有没有vscode上的编译环境代码啊哥

**Reply by teacup418** [reply]
去cs50.dev弄，新手就不要自己配cs50的环境了。除非你很熟悉docker，可以装好docker desktop后pull cs50/codespace，或者pip install check50。不然你就算头文件从github找了，复制到本地msys2的头文件库里面，代码写出来了也跑不了测试，无法提交。

**Comment by Dylan520520** [comment]
这个证书有啥用吗

**Comment by lng205** [comment]
分享习题及感想，希望为自学CS的同学提供参考：
https://github.com/lng205/CS50x

**Reply by lng205** [reply]
做到Week 8才发现mancuoj大佬的分享，所以没怎么参考。
总结了自己的收获，希望能提供一些不同的思路。
欢迎对习题或数字IC感兴趣的朋友来交流。

**Reply by zyxztmbs** [reply]
为什么我的作业和lab都打不开

**Reply by lng205** [reply]
刚刚试了一下打开课程网站和内部页面，[比如这个](https://cs50.harvard.edu/x/2022/psets/1/credit/)，没遇到问题。可以详细说说是哪个页面打不开吗。

**Reply by zyxztmbs** [reply]
我又试了一次，发现我不用梯子反而进去了，感谢回复😂

**Reply by lng205** [reply]
^ ^

**Reply by MisakiNemo** [reply]
我想请问一下，cs50x的提交时出现的ssh key之后要填写的是啥，已经再github创建了一个ssh key

**Reply by lng205** [reply]
> 我想请问一下，cs50x的提交时出现的ssh key之后要填写的是啥，已经再github创建了一个ssh key

[习题集](https://cs50.harvard.edu/x/2022/psets/1/)一开始有提到这个[教程页](https://cs50.readthedocs.io/github/#ssh)，按里面的步骤逐步操作就可以了

**Reply by Chononut** [reply]
vscode上的编译环境怎么调啊

**Reply by F1Justin** [reply]
我想层主的本意是好的，可是这个行为已经 VIOLATE CS50 的 ACADEMIC HONESTY POLICY 了。
规则https://cs50.harvard.edu/x/2025/honesty/中有这么一条：
Providing or making available solutions to assessments to anyone, whether a past, present, or prospective future student.
（

**Comment by fafa-chLoe** [comment]
打开这三个网页都是一片空白是什么情况TAT
校园网和流量都试过

**Reply by fafa-chLoe** [reply]
已解决 要挂梯子
之前不行是因为梯子有问题

**Reply by Jeremy1111111111111** [reply]
你好，想问下梯子是啥啊，我很多视频都看不了

**Reply by teacup418** [reply]
梯子以前指VPN，现在一般指代理服务器（proxy），比较常见的有clash和v2rayn/ng，软件内一般有机场主（代理服务器）的推广，随便挑一个价格合适的买就行了。clash和v2ray在github下载就可以，还有不懂的话再问

**Reply by zh-smiliue** [reply]
梯子：https://b.msclm.net/#/register?code=el7dimGa

**Comment by gaivrt** [comment]
这个课对我就相当于CS61A对博主，真的打开新世界大门

**Comment by Jeremy1111111111111** [comment]
梯子是什么啊，我很多课程都不能看

**Reply by Baymax-Rob** [reply]
这些课程大多数都在b站可以找到，搜索关键字：例如CS50x。

**Reply by Baymax-Rob** [reply]
分享CS50X-2023题目习题集，实验，目前学习和更新到week1,给喜欢学习的同学提供参考，顺手给个star： https://github.com/Baymax-Rob/Life-long-learning

**Reply by teacup418** [reply]
看这篇文章。
文章：https://digital-garden.top/pro/internet/magic/
密码：proxy

**Reply by zh-smiliue** [reply]
梯子：https://b.msclm.net/#/register?code=el7dimGa

**Comment by Baymax-Rob** [comment]
分享CS50X-2023题目习题集，实验，目前学习和更新到week1,给喜欢学习的同学提供参考，顺手给个star： https://github.com/Baymax-Rob/Life-long-learning

**Comment by shinehhht** [comment]
这个problem set怎么弄啊，没看懂教程在说啥，有没有新手指引什么的啊，谢谢谢谢谢谢

**Reply by Postedism** [reply]
可以看week1：problem set1里的 what to do，用网页翻译可以了解是要求你登录注册，之后一步步操作就行

**Reply by teacup418** [reply]
简单来说就是……你要开个github账号，然后登陆codespace，然后创建一个身份验证方便提交代码，然后用vscode写代码、调试，调试完检测一下语言风格是否具有易读性，然后提交给cs50的自定检查机器人，如果没问题的话就代表你成功的交了一次作业啦

**Comment by D-shuoshuo** [comment]
edx上有个CS50X

**Comment by Bright-Wong** [comment]
b站有 [2023 最新版](https://b23.tv/k6P1ThR) 了

**Comment by LiviaSun** [comment]
week0里里面的第三部上传.sb3文件是什么意思呀，是要上传什么呢

Go to submit.cs50.io/upload/cs50/problems/2022/x/scratch. Click “Choose File” and choose your .sb3 file. Click Submit.

**Reply by fish-404** [reply]
scratch的源文件

**Comment by xiechen274** [comment]
课程的setion 和shorts的内容和lecture有什么差吗

**Reply by JixunWu** [reply]
如果你对该周课的内容不熟，可以先看section再看lecture。如果你在听lecture的时候有不懂的地方可以看shorts（部分知识点更细致的讲解，算是对lecture对延伸)。要是你觉得自己掌握的比较好，只看lecture就行。复习或者不懂的也可以通读一遍notes。

**Comment by teacup418** [comment]
有没有Java课程的推荐？

**Comment by kathihalib** [comment]
795230802 QQ群，想从x开始学，然后把cs50和cs61系列学了

**Comment by teacup418** [comment]
拉了个群，点击链接加入QQ频道[Top数字花园](https://pd.qq.com/s/h4wshbnjy)

**Reply by teacup418** [reply]
关于梯子技术支持以及作业提交相关问题也可以问捏。

**Comment by lotusdezcliff** [comment]
這門課也沒想像中簡單，作為零基礎的小白每天還是要花上不少時間學習，但David Malan的教學真的太棒了。

**Reply by KnDellx956669** [reply]
同意你的看法

**Comment by Sikianri** [comment]
学完CS50 回来评论一下，首先这门课的推荐总学时是100-200小时（我用番茄钟记录下的学时也在这个范围内），其次这门课是用C入门的计算机导论课程，从课程安排就可以看出，Week1-Week5都是围绕C去讲解一些计算机基础概念，Week6接触Python，之后就很快的切换到SQL，HTML+CSS+JS，Flask，用3周时间迅速的带你从前端到后端过一遍。
    个人认为这么课还是很好的，无愧于在许多课程排行榜排第一，因为戴维马兰真是一名传奇教师，我打开教学视频第一眼就被他吸引了，能把编程课上出美国大片的既视感，就是这门课最大的优点（这对于ADHD的我非常重要，就像是网课版的《编码》），其次课程讲解清晰，Problem 提示循循善诱，但不失难度，很多地方自己做不出还是要向gpt4询问（私人助教），个人最喜欢的是Week5 C的大作业，自己单独写一个散列函数，并和教职员工的算法比速度，写了好几天，最后比staff的算法慢1-2倍，但真的很开心。
    其中最难的地方可能是Week8, HTML+CSS+JS 一股脑教给你，虽然都是些很基础的东西，但写到最后一个Problem人还是蒙的，不得已先跳到FreeCodeCamp再过了一遍html+css 基础才回来做。
    总之这门课非常有趣，我也非常感谢戴维马兰和CS50的教职员工，因为这门课实在是太棒了，很感谢他们能做出这么好的课。

    如果你想试试C入门的计算机导论课，或者当你学完CS61a想看看C，就选CS50吧（搭配K&R C效果更佳）

**Reply by Kettenkrado** [reply]
same adhd and same thought on this class. 感谢分享！

**Reply by skycccai** [reply]
感谢分享

**Comment by Andy-xiaokang** [comment]
## CS50的小结
### 印象深刻之处
* cs50p学到一半想起来过来评论，我觉得最难的是**tideman**那个问题，选举的问题层层递进，越来越难，最后的递归算法判断是否有环真的难，而且用数组来表示逻辑关系也很绕，其他部分还能自己完成，唯独那个递归算法死也想不到。我在查看medium别人的递归算法之前一直只想到了用循环去判定，结果总是测试不通过，最后在欣赏了别人的思路后“借鉴”了一下。
* 最欣慰的是**speller** hash table那个问题我分成了26\*26\*26+1个数组链表，其中的+1 表示存储第0个非正常链表（`strlen(word) < 3`）(`word[1] == '\'' || word[2] == '\''`)。所以load的速度比staff给出的速度还要快很多，但是find的速度却慢了很多，最后还是慢了很多。
* **recover**那个问题也是令我印象深刻，真的是能让人感到C语言操纵内存的魅力，对指针和文件的读写在实际问题中加深了很深刻的理解，最后还从这个实验里看了50张Harvard的校园照片，指针，内存那部分lab 和 pset 的debug一定会有`core dump, segmentation fault`, 因为太容易犯这个错误了。
* 老师对指针和字符串那部分的讲解真是让人拍案叫绝，我从来没有如此清晰地理解过指针和字符串（NUL和NULL的区别等等）。虽然我只做了week1-week7的课后练习题，但收获满满。
* 实在有太多印象深刻的了，写不完。
### 对David malan 和CS50的印象
* David malan 真是恩师，真的太精彩了，他没有一开始就给出最好最正确的代码，而是从新手的角度出发，把几乎所有问题都呈现，然后一步步抽象升华，一步步带你解决，最后展现出一个高手版本的代码，但是却给人一种自然以及我也可以写出这样的代码的感觉，告诉你为什么这样设计，给出详细的理由。比如通过对字符串的增添操作那里引人链表，以及展现链表灵活性那里真的惊艳，还有太多太多的细节我写不下。
* 试问有哪个老师能激情四射的给你讲授两个半小时高强度的课程中间只休息10min，一边讲课一边擦汗，不厌其烦的展示自己打磨过的最佳的课程还细心回答所有同学提出来的所有问题，每每想到如此，我都觉得David malan真是我见过的最优秀的老师，而且他的发音好标准准确啊，我也算是体会到了flyingpig说的不学好都感到惭愧的那种感觉。
* 整个课程的上课交互，演示就像看电影一样，给人一种这TM就是Harvard顶级学府的魅力吗？课后作业难度很高的都有助教录制专门的讲解视屏，带你更好的理解问题，lab和pset的难度层层递进，而且都是从具体问题出发，那些问题都有具体的背景，都很有意思，用来当英语阅读练习都不错，而不是像国内一样自己编写那种教条式的题目，而且从课后练习量来讲也可以看出国外计算机教育真的极其重视实践的特点，和照本宣科这个词沾不上一点关系
* 这是一门寓教于乐但不只有乐的课程，他的难度远高于cs50P, 也能体会到很多像David在CS50P中说的“绞尽脑汁后获得解决一个问题所带来的欣喜感”

最后感谢一下这位贡献者mancuoj，我觉得这对我是一个好的开始，来自一个曾经被教的毫无兴趣的人甚至怀疑自己是否适合计算机的人的评价。

**Comment by taylover2016** [comment]
20230727更新，CS50要有一门专门的课程讲SQL了！泪目

**Reply by lotusdezcliff** [reply]
已經有了~
https://cs50.harvard.edu/sql/2024/

**Comment by TephrocactusMYC** [comment]
对纯粹的初学者来说可以考虑花一个学期来做，但对于有一定基础的学生来说可以直接看notes然后做作业了，个人认为tideman最难。

**Reply by TephrocactusMYC** [reply]
我只认真做了C部分，后面前端那些乱七八糟的水了水，没有深入。

**Comment by Zephyr-Yara** [comment]
大佬救救，#include<cs50.h>这个库在哪里下载啊

**Reply by Andy-xiaokang** [reply]
直接在网页版的vs code实验环境里添加，不用下载。好像网页上有提到过这个库

**Reply by Zephyr-Yara** [reply]
谢谢佬！已经弄好了。

**Comment by teacup418** [comment]
要么用官方云环境编程，要么在github搜索cs50.h，然后把头文件下载后粘贴到本地的库。

从 Windows 版邮件发送

发件人: Mirrow-Rain
发送时间: 2023年8月14日 17:10
收件人: PKUFlyingPig/cs-self-learning
抄送: mind-dance; Comment
主题: Re: [PKUFlyingPig/cs-self-learning] Harvard CS50: This is CS50x - CS自学指南 (Discussion #87)

大佬救救，#include<cs50.h>这个库在哪里下载啊

**Comment by oliverxt** [comment]
看shorts之前记得调小音量

**Comment by Rimor4** [comment]
请问想学做游戏比如unity的话，推荐学完cs50x后继续学后续的cs50 Introduction to Game Development吗

**Comment by RandABC** [comment]
答案在哪看啊大家知道吗，课后习题的答案

**Reply by Rimor4** [reply]
上面【资源汇总】有大佬的github，代码里有每周的problem set

**Comment by imsubei** [comment]
**CS50有交流的伙伴吗，交流群：540468803 可以一起讨论一下**

**Comment by PeteHui** [comment]
一开始以为这门课面向所有学生公选，想当然认为这门课的作业很简单，自己做的时候往往需要在电脑跟前苦熬好几个小时，还以为自己太笨了，是不是其他人会觉得很简单。看到其他朋友的感受后，顿时觉得：“I am not alone.”

**Reply by Link948** [reply]
不简单🫡

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by Kettenkrado** [comment]
官方有建一个Discord社群，里面有很多在学/学完了的人和讨论的留存聊天记录，我今天开始上这门课，并且和一些在上这门课的人组成了一个小的学习群（进度大概都是刚开始的样子），下面贴上官方链接和我的discord链接，如果你想要一起讨论学习欢迎加入。
官方：https://discord.gg/cs50
我的用户名：kettenkrado

**Reply by Liu-Hong-Wei** [reply]
嗨，我也是三月初开始的，可以一起讨论嘛，discord上请求加你啦

**Comment by nooblesone** [comment]
这个用来学英语挺好的（

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by Miokasa** [comment]
可以更新2024吗

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by Chononut** [comment]
初学计算机 以考研为目的学这个合适吗？刚听了前面的第一节课，感觉用了太多时间去比喻二进制了。。 然后看视频目录对比了一下C语言教材，是不是教材的有些东西没有讲啊，现在有点迷茫，不知道该不该继续听这个。

**Reply by Grignard-reagent** [reply]
不合适。它和408差距就像3b1b的线代与你期末考的线代难度差不多大。且本课极为基础，可视作预科打基础提兴趣之用，考研时间来不及。建议自行寻找辅导班教辅网课备战考研。

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by Chononut** [comment]
大佬们，vscode上的编译环境有没有调试好的代码啊？在网上很多种方法都试了还调不出来

**Reply by fish-404** [reply]
可以直接用 offline with docker 的方案，本地配一个环境出来

**Comment by fish-404** [comment]
codesapce 打不开的情况可以使用 offline with docker 的方案，这个方案你可以得到一个 codespace 的环境，不过 `check50` 和 `submit50` 可能还是无法使用，这时候可以到你的 Github 有个 me50 的仓库，在这里按照 slug 创建分支，提交每个 Problem Set 的代码，就可以在课程提交中看到了，那边会有 check50 的结果

**Reply by fish-404** [reply]
`check50` 也可以直接使用 `check50 --local` 或者 `check50 --offline`

**Comment by ChenjunxiNB** [comment]
编程入门根本学不完，请问一下如果只能听一半的话建议学什么?

**Comment by chiueryue** [comment]
cs50的习题难度对于初学者来说好像并不友好   打算先学下一个

**Comment by Flow-sandyu** [comment]
习题确实不简单，但如果是新手的话，值得做。很好的检验自己学习掌握程度，锻炼写代码能力

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by gaojinyuhhhhhh** [comment]
大佬们，这个网站为什么打不开

**Comment by LINYV0719** [comment]
我最开始尝试登录codespace，但是无论如何都登不上去，后来我试着换了一个用谷歌邮箱注册的github账号，最后成功登录

**Reply by LINYV0719** [reply]
为遇到了和我一样问题的人提供参考

**Reply by bing1117** [reply]
我今年一月学的时候还可以登陆codespace的，我还提交过作业，9月要登陆codespace老是offline，我甚至重新创建了一个账号登陆还是offline

**Reply by LINYV0719** [reply]
是梯子的问题吗，我这里选择打开我的VSCODE，完全没有遇到offline的问题

**Reply by Chozzc** [reply]
用vscode登陆试试

**Reply by bing1117** [reply]
对是梯子的问题，最后换了个梯子，直接好了

**Comment by Chozzc** [comment]
不想翻墙的可以去这个网站，里面资源还是蛮多的https://www.learncs.site/

**Comment by NyxScheppen** [comment]
感觉新手学大概得40h起步...20h才学到week5的人如是说

**Comment by 1ceshadow** [comment]
挂不挂梯子都一直显示setting up your codespace😭

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by ME3666** [comment]
哪位大佬有shorts的中文版

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by Eagle233Fake** [comment]
妈的，这么吊的课，学一半都是对他的不尊重

**Comment by juanniaoxx** [comment]
C语言学习强烈推荐南大的CPL，🐜开放了OJ账户的申请！

**Comment by Chononut** [comment]
在哪里申请哇佬&nbsp;



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年12月10日(周二) 下午2:41
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [PKUFlyingPig/cs-self-learning] Harvard CS50: This is CS50x - CS自学指南 (Discussion #87)




 
C语言学习强烈推荐南大的CPL，🐜开放了OJ账户的申请！

**Reply by juanniaoxx** [reply]
视频的话b站王老师和🐜老师都开放了。OJ账户可以https://www.bilibili.com/video/BV1SWtQexEEo?spm_id_from=333.788.videopod.sections&vd_source=6a2e8fe6292e2cc33457c6e34734309a,老师在评论里面写了！

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by whynot-37** [comment]
有人知道作业在网站哪里找到吗？左边黑色框和课程视频下方都没看见啊。。。

**Reply by teacup418** [reply]
https://cs50.harvard.edu/x/2025/psets/1/

**Reply by whynot-37** [reply]
谢谢大佬

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by ITstudent111** [comment]
Hello Guys,

This is Fahad, I hope you all are doing well, I am new to CS50 course, I have some questions ,can I post some questions here ?

Thanks

**Comment by xielinhegel** [comment]
2024年的刚开始看，2025年的就出来了。听哪个好呢/(ㄒoㄒ)/~~

**Reply by Carter998** [reply]
个人觉得2024版更好一些

**Reply by teacup418** [reply]
计算机领域跟最新的。况且2024的作业你也做不了了，都会统一迁移到2025的。

**Comment by Javien-Li** [comment]
这是我在大一一开始就认识到的一门课程,算是我的启蒙吧.马兰教授的幽默,真是让人难忘

**Comment by yf-wang711** [comment]
想问一下，创建codespace的时候整个codespace的页面已经出来了，但是却无法创建文件和文件夹是为什么，还有edx账号是必须的吗

**Reply by yf-wang711** [reply]
靠个人感觉好像是梯子太卡了加载太慢

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by csy01-xy** [comment]
所以请问这种网络公开课是最好跟最新版本还是说都行？因为老版本有精翻而最新版的基本就只有ai翻译了。我的英语水平还是没那么好，直接看生肉有些词听着还是略感吃力，想请教一下各位前辈。

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by Draconis011** [comment]
有没有佬指导一下怎么提交作业呀 新手有点懵😭

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！

**Comment by Sprallow** [comment]
怎么做到20小时就学完的？我这里起码要70小时啊

**Comment by PeteHui** [comment]
您的邮件我已经收到，谢谢！
